Kube Play - pass arguments to build

ygalblum opened the pull request:
Create a buildah SystemContext from the existing cli arguments
Pass the SystemContext to the build
Add system test

#### Does this PR introduce a user-facing change?
No

```release-note
None
```

Resolves: #20890

rhatdan commented:
/approve
LGTM

edsantiago commented on test/system/700-play.bats:
```diff
@@ -927,3 +928,42 @@ spec:
+
+    run_podman stop -a -t 0
+    run_podman pod rm -t 0 -f test_pod
+    run_podman rmi -f userimage:latest
```
Please also rmi the other image
```diff
diff --git a/test/system/700-play.bats b/test/system/700-play.bats
index a5284ae3b..5eaeb8568 100644
--- a/test/system/700-play.bats
+++ b/test/system/700-play.bats
@@ -967,3 +967,3 @@ _EOF
     run_podman pod rm -t 0 -f test_pod
-    run_podman rmi -f userimage:latest
+    run_podman rmi -f userimage:latest $from_image
 }
```

ygalblum replied:
Done

ygalblum commented:
@edsantiago maybe you help me find what I'm doing wrong here. The test I added always fails when running all the tests (locally or in the CI) and sometimes when running it separately locally. It always fails in the `run_podman login` line: https://github.com/containers/podman/pull/20966/files#diff-d3acce6fc9725bf464aafb70ec081bf119159d098c06c76204cb18559e2bbfacR947
I copied these lines from other tests that I've never seen failed at this point. Not sure what's wrong here. WDYT?

edsantiago commented:
I stared at it a long time this morning and could not figure it out. I'm on PTO today, but I just stared at it some more, and still can't understand it. 

Could you try TEMPORARILY adding `sleep 2` between `start_registry` and the login? `start_registry` tries hard to make sure that the registry is running, but maybe that code is broken?

I will look at it again Tuesday morning with a fresh brain.

edsantiago commented:
Don't bother with the `sleep`. That's not it. There's a sequencing problem; the test failure only happens on a full bats run. I've tried, unsuccessfully, to reproduce it with a reduced set of tests. No luck. And it's bedtime.

ygalblum commented:
@edsantiago thanks for looking into this.

I was able to reproduce this failure on my machine (Fedora 38) running `./hack/bats --root 700:"private registry"`. The test passes when running with `--rootless`. 

I have tried adding a `sleep 5` between `start_registry` and the `login`, but got the same results:
```
700-play.bats
 ✗ [700] podman play --build private registry
   (from function `bail-now' in file test/system/helpers.bash, line 227,
    from function `die' in file test/system/helpers.bash, line 790,
    from function `run_podman' in file test/system/helpers.bash, line 415,
    in test file test/system/700-play.bats, line 948)
     `run_podman login --authfile=$authfile \' failed
   
   [11:25:06.810377103] # /home/ygal/work/podman/bin/podman rm -t 0 --all --force --ignore
   
   [11:25:06.846237978] # /home/ygal/work/podman/bin/podman ps --all --external --format {{.ID}} {{.Names}}
   
   [11:25:06.879006955] # /home/ygal/work/podman/bin/podman images --all --format {{.Repository}}:{{.Tag}} {{.ID}}
   [11:25:06.906438849] quay.io/libpod/testimage:20221018 f5a99120db64
   
   [11:25:06.934846841] # /home/ygal/work/podman/bin/podman image exists quay.io/libpod/registry:2.8
   [11:25:06.968021367] [ rc=1 ]
   # skopeo copy --all oci-archive:/tmp/podman-systest-imagecache-0/quay.io--libpod--registry--2.8.tar containers-storage:[vfs@/tmp/podman-bats-registry.LPvrrd/root+/tmp/podman-bats-registry.LPvrrd/runroot]quay.io/libpod/registry:2.8
   Getting image source signatures
   Copying blob sha256:40e059520d199e1a1a259089077f2a0c879951c9a4540490bad3a0d7714c6ae7
   Copying blob sha256:3b494d572400cf4b26beff6a89af57b70fbc1d7a2c1aa7a0d7b67c756bdaed61
   Copying blob sha256:88ce0e919729180f43f37c386397ee7df6154e172f4f104c7196572d2eb88919
   Copying blob sha256:0fec280ccc88586d94f581861ff02b7e9c16fc128e61be7315365e18efa9994f
   Copying blob sha256:d637f99b74411f98a3d6a4e4f9a496eac649828eac5253d596eac97128af1aaf
   Copying config sha256:d3241e050fc981849fc396c054aa87da6b1e2d748ee46cd3e71bf89050881631
   Writing manifest to image destination
   ..+........+....+.....................+..+..........+..+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*...+.....+.+............+..+.+......+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*.......+...............+............+..........+..+....+.....................+........+.............+........+.+...+........+..........+........+....+..+........................+.+......+.........+...+...+........+.+........+...+...+.+......+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
   ..........+.+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*...+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*.+...+.....+...+.......+...+.........+...........+........................................+..+...+.......+..+...+.+......+...........+...+......+.........+.......+..+...+...+.+....................+....+.....+.....................................+..+.....................+.........+.+...........+...+...........................+...+..................+...................+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
   -----
   
   [11:25:08.314806178] # /home/ygal/work/podman/bin/podman --storage-driver vfs --root /tmp/podman-bats-registry.LPvrrd/root --runroot /tmp/podman-bats-registry.LPvrrd/runroot run -d -p 127.0.0.1:42706:5000 --name registry -v /tmp/podman-bats-registry.LPvrrd/auth:/auth:Z -e REGISTRY_AUTH=htpasswd -e REGISTRY_AUTH_HTPASSWD_REALM=Registry Realm -e REGISTRY_AUTH_HTPASSWD_PATH=/auth/htpasswd -e REGISTRY_HTTP_TLS_CERTIFICATE=/auth/domain.crt -e REGISTRY_HTTP_TLS_KEY=/auth/domain.key quay.io/libpod/registry:2.8
   [11:25:08.545440724] dd5761f9fe910c76ee24ae274187e7f77ecc6a3a3f087a5a9dcf732a0545aac3
   
   [11:25:08.582389620] # /home/ygal/work/podman/bin/podman logs dd5761f9fe910c76ee24ae274187e7f77ecc6a3a3f087a5a9dcf732a0545aac3
   [11:25:08.643875159] time="2023-12-12T09:25:08.549260359Z" level=warning msg="No HTTP secret provided - generated random secret. This may cause problems with uploads if multiple registries are behind a load-balancer. To provide a shared secret, fill in http.secret in the configuration file or set the REGISTRY_HTTP_SECRET environment variable." go.version=go1.16.15 instance.id=ee8c8d40-8b07-4bce-a89e-7a5923ef4bc2 service=registry version="v2.8.1+unknown"
   time="2023-12-12T09:25:08.549294722Z" level=info msg="redis not configured" go.version=go1.16.15 instance.id=ee8c8d40-8b07-4bce-a89e-7a5923ef4bc2 service=registry version="v2.8.1+unknown"
   time="2023-12-12T09:25:08.549347909Z" level=info msg="Starting upload purge in 48m0s" go.version=go1.16.15 instance.id=ee8c8d40-8b07-4bce-a89e-7a5923ef4bc2 service=registry version="v2.8.1+unknown"
   time="2023-12-12T09:25:08.556421626Z" level=info msg="using inmemory blob descriptor cache" go.version=go1.16.15 instance.id=ee8c8d40-8b07-4bce-a89e-7a5923ef4bc2 service=registry version="v2.8.1+unknown"
   time="2023-12-12T09:25:08.556720307Z" level=info msg="restricting TLS version to tls1.2 or higher" go.version=go1.16.15 instance.id=ee8c8d40-8b07-4bce-a89e-7a5923ef4bc2 service=registry version="v2.8.1+unknown"
   time="2023-12-12T09:25:08.556750411Z" level=info msg="restricting TLS cipher suites to: TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_AES_128_GCM_SHA256,TLS_CHACHA20_POLY1305_SHA256,TLS_AES_256_GCM_SHA384" go.version=go1.16.15 instance.id=ee8c8d40-8b07-4bce-a89e-7a5923ef4bc2 service=registry version="v2.8.1+unknown"
   time="2023-12-12T09:25:08.557035925Z" level=info msg="listening on [::]:5000, tls" go.version=go1.16.15 instance.id=ee8c8d40-8b07-4bce-a89e-7a5923ef4bc2 service=registry version="v2.8.1+unknown"
   
   [11:25:13.660115800] # /home/ygal/work/podman/bin/podman login --authfile=/tmp/podman_bats.C2TSLm/authfile.json --tls-verify=false --username userRR8g --password pwLdgyt2JWbYWlty2 localhost:42706
   [11:25:26.026492376] Error: authenticating creds for "localhost:42706": pinging container registry localhost:42706: Get "http://localhost:42706/v2/": dial tcp [::1]:42706: connect: connection refused
   [11:25:26.030380734] [ rc=125 (** EXPECTED 0 **) ]
   #/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
   #| FAIL: exit code is 125; expected 0
   #\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   
   [11:25:26.038859914] # /home/ygal/work/podman/bin/podman pod rm -t 0 -f -a
   
   [11:25:26.063366307] # /home/ygal/work/podman/bin/podman rm -t 0 -f -a
   
   [11:25:26.086404981] # /home/ygal/work/podman/bin/podman image list --format {{.ID}} {{.Repository}}
   [11:25:26.109755169] f5a99120db64 quay.io/libpod/testimage
   # [teardown]
   
   [11:25:26.111098191] # /home/ygal/work/podman/bin/podman pod rm -t 0 --all --force --ignore
   
   [11:25:26.131669940] # /home/ygal/work/podman/bin/podman rm -t 0 --all --force --ignore
   
   [11:25:26.152118892] # /home/ygal/work/podman/bin/podman network prune --force
   
   [11:25:26.180125589] # /home/ygal/work/podman/bin/podman volume rm -a -f

1 test, 1 failure
```

ygalblum commented:
Having said that, while the failure on my machine is consistent, when I run the test on a Fedora Rawhide VM, the test passes

edsantiago commented:
That's weird. On my F39 laptop, plain `hack/bats 700:private` succeeds, both root and rootless, even in a loop. It only fails if I run `hack/bats` plain, i.e. so it runs the entire test suite. I'm continuing to look into it this morning.

edsantiago commented:
Found it. It may take me a few hours before I can get to working on a fix, but I'll do so tonight.

edsantiago commented:
Fix is in #20998, which is unlikely to merge soon or as-is, but feel free to cherry-pick it so you can at least run your PR through full CI.

edsantiago commented on test/system/700-play.bats:
```diff
@@ -927,3 +928,42 @@ spec:
+    run_podman image rm --ignore $from_image
+
+    _write_test_yaml command=id image=userimage
+    run_podman 125 play kube --build --start=false $PODMAN_TMPDIR/test.yaml
```
Please check error messages every single time:
```diff
diff --git a/test/system/700-play.bats b/test/system/700-play.bats
index ffe692103..f50acd9e6 100644
--- a/test/system/700-play.bats
+++ b/test/system/700-play.bats
@@ -967,4 +967,7 @@ _EOF
     _write_test_yaml command=id image=userimage
     run_podman 125 play kube --build --start=false $PODMAN_TMPDIR/test.yaml
+    assert "$output" = "Error: short-name resolution enforced but cannot prompt without a TTY" \
+           "I do not think this is the error you were expecting here"
+
     run_podman play kube --replace --context-dir=$PODMAN_TMPDIR --tls-verify=false --authfile=$authfile --build --start=false $PODMAN_TMPDIR/test.yaml
     run_podman inspect --format "{{ .Config.User }}" test_pod-test
```

ygalblum replied:
Done

edsantiago commented:
Phew. #20998 is ready. If it hasn't merged by the time you start your day, please cherrypick it. And please look at my two comments above.

ygalblum commented:
> Phew. #20998 is ready. If it hasn't merged by the time you start your day, please cherrypick it. And please look at my two comments above.

Thanks a lot. I've rebased and added your comments. I've also added a `podman image pull $IMAGE` at the start to avoid failing when running a single test.

edsantiago commented:
>I've also added a podman image pull $IMAGE

Eeek! No, please no! `setup()` guarantees that this image will be present.Running `pull` is just a recipe for flake disaster.

>assert "Error: short-name resolution enforced but cannot prompt without a TTY

So, I think this is going to be less easy than I'd hoped. That's the message on my laptop. On CI, it's different.

ygalblum commented:
@edsantiago I've removed both lines. 
I can do without checking the error output.
I'll try to find time to understand why the image is not being pre-fetched when running a single test

edsantiago commented:
> I can do without checking the error output.

No, what I was trying to say is that it is imperative to check error output, I just don't know how. Maybe `assert "$output" =~ "pattern1|pattern2"`

It is super-critical to check error output because we've had way, way too many instances of
```
   run_podman 125  something-with-a-typo
   [ fails, but not for the expected reason]
```
Or sometimes not a typo, just an unexpected condition.

> I'll try to find time to understand why the image is not being pre-fetched when running a single test

That's because of the missing `rmi $something` that I had you add. TL;DR `setup()` takes pains to remove unwanted images, but in the one-test-only case where it leaves `$IMAGE` _and_ the same image with a different name, it fails. This is not important enough to fix: the code would be much more complicated and unmaintainable than any possible value. Unfortunately, in the development cycle with lots of failed iterations -- like your case right now -- it sucks. And I'm sorry about that.

ygalblum commented:
@edsantiago I've updated the test to check either cases

rhatdan commented:
@edsantiago PTAL

edsantiago commented:
Test LGTM. Your PR exposed some serious instability in our test framework. Thank you for giving me time to fix it.

(The non-test code, I leave for others to review)

rhatdan commented:
/lgtm
